pbm/restore/physical.go

```diff
@@ -0,0 +1,885 @@
+package restore
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	slog "log"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"golang.org/x/mod/semver"
+
+	"github.com/percona/percona-backup-mongodb/pbm"
+	"github.com/percona/percona-backup-mongodb/pbm/log"
+	"github.com/percona/percona-backup-mongodb/pbm/storage"
+)
+
+const (
+	defaultRSdbpath   = "/data/db"
+	defaultCSRSdbpath = "/data/configdb"
+
+	mongofslock = "mongod.lock"
+
+	defaultPort = 27017
+)
+
+type PhysRestore struct {
+	cn     *pbm.PBM
+	node   *pbm.Node
+	dbpath string
+	port   int
+	// an ephemeral port to restart mongod on during the restore
+	tmpPort string
+	rsConf  *pbm.RSConfig // original replset config
+
+	name     string
+	opid     string
+	nodeInfo *pbm.NodeInfo
+	stg      storage.Storage
+	bcp      *pbm.BackupMeta
+	// Shards to participate in restore. Num of shards in bcp could
+	// be less than in the cluster and this is ok.
+	//
+	// Only the restore leader would have this info.
+	shards []pbm.Shard
+
+	stgpath string
+
+	stopHB chan struct{}
+
+	log *log.Event
+}
+
+func NewPhysical(cn *pbm.PBM, node *pbm.Node, inf *pbm.NodeInfo) (*PhysRestore, error) {
+	opts, err := node.GetOpts()
+	if err != nil {
+		return nil, errors.Wrap(err, "get mongo options")
+	}
+	p := opts.Storage.DBpath
+	if p == "" {
+		switch inf.ReplsetRole() {
+		case pbm.RoleConfigSrv:
+			p = defaultCSRSdbpath
+		default:
+			p = defaultRSdbpath
+		}
+	}
+	if opts.Net.Port == 0 {
+		opts.Net.Port = defaultPort
```

Comment:
depends on rs role default port is different

https://docs.mongodb.com/manual/reference/default-mongodb-port/

Reply:
Nice catch.
Though in fact, `opts.Net.Port` now is only used to [calculate tmp port](https://github.com/percona/percona-backup-mongodb/pull/638/files/6bdbcf9a420193b62523e93fba2d35fcfc8c248a#diff-d99d228622352026338b852f0e347ffa7e403639e5460fbb6205ec955a98b197R100) so it doesn't matter. I was saving in order to restart mongod in prod env after the restore. But since we leave the final restart to the user, I will rid of `opts.Net.Port` along with `PhysRestore.port` in the next release.